x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -413,22 +412,25 @@ public void restoreShard(
                 CcrRetentionLeases.RETENTION_LEASE_RENEW_INTERVAL_SETTING.get(store.indexSettings().getNodeSettings()),
                 Ccr.CCR_THREAD_POOL_NAME
             );
+            // Some tests depend on closing session before cancelling retention lease renewal
             toClose.add(() -> {
                 logger.trace("{} canceling background renewal of retention lease [{}] at the end of restore", shardId, retentionLeaseId);
                 renewable.cancel();
             });
             // TODO: There should be some local timeout. And if the remote cluster returns an unknown session
             // response, we should be able to retry by creating a new session.
-            final RestoreSession restoreSession = openSession(metadata.name(), remoteClient, leaderShardId, shardId, recoveryState);
-            toClose.addFirst(restoreSession); // Some tests depend on closing session before cancelling retention lease renewal
-            restoreSession.restoreFiles(store, restoreListener.delegateFailureAndWrap((l, v) -> {
-                logger.trace("[{}] completed CCR restore", shardId);
-                updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, shardId.getIndex());
-                l.onResponse(null);
-            }));
-        } catch (Exception e) {
-            restoreListener.onFailure(e);
-        }
+            ActionListener<RestoreSession> sessionListener = restoreListener1.delegateFailureAndWrap((l1, restoreSession) -> {
+                restoreSession.restoreFiles(
+                    store,
+                    l1.delegateResponse((listener1, e) -> restoreSession.close(listener1)).delegateFailureAndWrap((l2, v) -> {
+                        logger.trace("[{}] completed CCR restore", shardId);
+                        updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, shardId.getIndex());
```

Comment:
Ugh right so it is. I think that's enough async steps in a row that it'd make sense to do this with a sequence of `SubscribableListener` things instead of nesting all these listeners directly. I'm ok with deferring some of the changes to future PRs but would it make sense to start introducing `SubscribableListener` things here already? Note that `SubscribableListener#addTimeout` exists, you wouldn't need to use `ListenerTimeouts` if so.

Reply:
I did try to make also `updateMappings` async and use `SubscribableListener` (https://github.com/elastic/elasticsearch/pull/98035/commits/4c29bfbd398291f8683200ab04d920d86ec0ea2f) but some test keeps failing on me. I think I miss some forking somewhere, but for now I'd rather keep that for a follow up. Also considering an async `updateMappings` would be where we could really use some `SubscribableListener`, I'd defer that too, if that's ok with you.